### PR TITLE
distinct descriptor mode

### PR DESCRIPTION
**Pull request by zscole**

### CODE REVIEW

File: onionbalance/hs_v3/params.py [hunk 1]
```diff
@@ -48,3 +48,7 @@
 # Misc parameters
 
 DEFAULT_LOG_LEVEL = os.environ.get('ONIONBALANCE_LOG_LEVEL', 'warning')
+
+
+MAX_INTRO_POINTS=10
```
Comment (asn-d6) [inline onionbalance/hs_v3/params.py:53]: Hmm, where did this 10 value come from? Can you consistently fit 10 intro points in a descriptor? If yes, let's change the name to `MAX_INTRO_POINTS_PER_DESCRIPTOR` and document it appropriately.

File: onionbalance/hs_v3/service.py [hunk 2]
```diff
@@ -263,6 +263,28 @@ def publish_descriptors(self):
         self._publish_descriptor(is_first_desc=True)
         self._publish_descriptor(is_first_desc=False)
 
+    
+    def _get_intros_for_distinct_desc(self, maxCount=10):
```
Comment (asn-d6) [inline onionbalance/hs_v3/service.py:267]: Nitpick: `maxCount` is camel cased, when the rest of the code is underscore-based. Let's make all new code use underscores please.

File: onionbalance/hs_v3/service.py [hunk 3]
```diff
@@ -263,6 +263,28 @@ def publish_descriptors(self):
...
+        # Get number of instances that contributed to final intro point list
+        n_instances = len(all_intros.intro_points)
+
+        final_intros = all_intros.choose(maxCount)
+
+        if (len(final_intros) == 0):
+            logger.info("Got no usable intro points from our instances. Delaying descriptor push...")
+            raise NotEnoughIntros
+
+        # logger.info("We got %d intros from %d instances. We want %d intros ourselves (got: %d)",
```
Comment (asn-d6) [inline onionbalance/hs_v3/service.py:282]: This log message got removed but it was useful. We should reinstall it somewhere.

File: onionbalance/hs_v3/service.py [hunk 4]
```diff
@@ -294,67 +316,102 @@ def _publish_descriptor(self, is_first_desc):
...
-            intro_points = self._get_intros_for_desc()
-        except NotEnoughIntros:
-            return
+        max_intro_points = params.MAX_INTRO_POINTS
 
-        # Derive blinding parameter
-        _, time_period_number = hashring.get_srv_and_time_period(is_first_desc)
-        blinding_param = my_onionbalance.consensus.get_blinding_param(self._get_identity_pubkey_bytes(),
-                                                                      time_period_number)
+        distinct_descriptors = params.DISTINCT_DESCRIPTORS
```
Comment (asn-d6) [inline onionbalance/hs_v3/service.py:321]: I like the idea of using distinct descriptors only when it's needed, so I really like the logic here. In this case, let's not make distinct descriptors be a global parameter, and let's just have this logic here to decide whether to enable this feature or not.

We don't need to expose this "distinct descriptor" feature to the user, and onionbalance can just intelligently decide whether to enable it or not. I like this.

File: onionbalance/hs_v3/service.py [hunk 5]
```diff
@@ -294,67 +316,102 @@ def _publish_descriptor(self, is_first_desc):
...
-        except hashring.EmptyHashRing:
-            logger.warning("Can't publish desc with no hash ring. Delaying...")
-            return
 
-        desc.set_last_publish_attempt_ts(datetime.datetime.utcnow())
-
-        logger.info("Uploading %s descriptor for %s to %s",
-                    "first" if is_first_desc else "second",
-                    self.onion_address, responsible_hsdirs)
+        if distinct_descriptors:
```
Comment (asn-d6) [inline onionbalance/hs_v3/service.py:326]: Hm. I don't quite understand the logic here. Can you please move this to a new function and document how this distinct descriptors feature should work?

File: onionbalance/hs_v3/service.py [hunk 6]
```diff
@@ -294,67 +316,102 @@ def _publish_descriptor(self, is_first_desc):
...
+                except NotEnoughIntros:
+                    return
+                logger.info("uploading a distinct descriptor")
+                try:
+                    desc = descriptor.OBDescriptor(self.onion_address, self.identity_priv_key,
+                                                   blinding_param, intro_points, is_first_desc)
+                except descriptor.BadDescriptor:
+                    return
+                self._upload_descriptor(my_onionbalance.controller.controller, desc, hsdir)
+        else:
```
Comment (asn-d6) [inline onionbalance/hs_v3/service.py:354]: So this is the non-distinct descriptor mode? is this identical to the previous functionality of Onionbalance?
In any case, let's move this also to its own function and document how it works.

**Review by asn-d6 — CHANGES_REQUESTED** [review]
Thanks for all the code!

I have a bunch of questions about your work:
1) How can I test that this feature works as intended? How did you test it yourself?
2) What's the logic behind the distinct descriptor mode? How does it work exactly? This should be a documentation inside the code (and also on the OBv3 docs but I can do that myself after I understand how it works).
3) How many introduction points can this now support at max?
4) Can you please make a unittest that demonstrates that the introduction points got distributed properly to the various descriptors?
5) Seems like the branch has conflicts with master, and also CI is broken.

Thanks again :)

**Comment by zscole** [comment]
Sorry for delayed response. Your patience is appreciated. Will have follow-up PRs submitted over the weekend. Excited to be contributing to this project!

**Comment by asn-d6** [comment]
Unfortunately I don't see this PR going anywhere. I close it. Feel free to reuse the code to do something good,.

**Comment by artembokhan** [comment]
Could anybody comment why the bakend limit is 20 in DD mode for v3?

ps. PR is working

**Comment by asn-d6** [comment]
> Could anybody comment why the bakend limit is 20 in DD mode for v3?
> 
> ps. PR is working

There is no DD mode in Onionbalance right now. This PR was left incomplete and far from done.

This code needs more work, and a decent final result can scale to any number of services.

**Pull request closed without merging**